include/at_cmd_interface.h

```diff
@@ -0,0 +1,90 @@
+/**
+ * @file at_cmd_interface.h
+ *
+ * @defgroup at_cmd_interface AT Command interface driver
+ *
+ * @{
+ *
+ * @brief Public APIs for the AT command interface driver.
+ */
+
+#ifndef ZEPHYR_INCLUDE_AT_CMD_INTERFACE_H_
+#define ZEPHYR_INCLUDE_AT_CMD_INTERFACE_H_
+/*
+ * Copyright (c) 2019 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: LicenseRef-BSD-5-Clause-Nordic
+ */
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+#include <zephyr/types.h>
+
+/**
+ * @brief AT command return codes
+ */
+enum at_cmd_rc {
+	AT_CMD_OK,
+	AT_CMD_ERROR,
+	AT_CMD_CMS,
+	AT_CMD_CME,
+	AT_CMD_NO_RETURN_CODE,
+};
+
+/**
+ * @typedefs at_cmd_interface_handler_t
+ * @brief Callback API when a message is received from the mode
+ *
+ * @param "char * response" Null terminated string containing the modem message
+ * @param "size_t response_len" Length of string
+ *
+ */
+typedef void (*at_cmd_handler_t)(char *response,
+				 size_t response_len);
+
+/**
+ * @brief Function to send an AT command to the modem
+ *
+ * @param cmd Pointer to null terminated AT command string
+ */
+enum at_cmd_rc at_cmd_write(const char *const cmd);
+
+/**
+ * @brief Function to send an AT command and receive response immediately
+ *
+ * For AT commands that will return a response immediately before the
+ * the command return code can be called by this function to get the response
+ * string in the supplied buffer if it's large enough. If the buffer is not
```

Comment:
So here is my conclusion after evaluating your proposals:

I can not trim the response because could potentially break the meaning of the information in the response.

If the internal buffer was static I could not point to it because I don't know when it would be reused for a new message. If the internal buffer was dynamic I could not point to it because then I would not know when to release it.

I could however remove the fallback, and always return NULL if the buffer is to small with an LOG_ERR message.

Reply:
I think this would be reasonable to log this as an error in case you do not want to trim the message.

Reply:
Ok